### PR TITLE
ENH: demux-paired

### PR DESCRIPTION
**Pull request by thermokarst**
The diff is a bit gnarly (esp. in the tests), it makes this PR look a lot worse than it is, IMO.

### CODE REVIEW

File: q2_cutadapt/_demux.py [hunk 1]
```diff
@@ -33,29 +34,61 @@ def run_command(cmd, verbose=True):
...
-        out_fp = per_sample_dir_fmt.sequences.path_maker(sample_id=sample_id,
-                                                         barcode_id=barcode_id,
-                                                         lane_number=1,
-                                                         read_number=1)
-        src = os.path.join(str(per_sample_dir_fmt), '%s.fastq.gz' % sample_id)
-        if os.path.isfile(src):
-            os.rename(src, str(out_fp))
+def _rename_files(seqs_dir_fmt, per_sample_dir_fmt, barcode_series):
+    # PAIRED-END
+    if isinstance(seqs_dir_fmt, MultiplexedPairedEndBarcodeInSequenceDirFmt):
```
Comment (ebolyen) [inline q2_cutadapt/_demux.py]: You could factor out this if-else branch to something like this:
```python
read_directions = [1]
if isinstance(...)
    read_directions.append(2)

for (sample_id, barcode_id) in barcode_series.iteritems():
    for read_direction in read_directions:
        out_fp = ...
        src = ...
        if os.path.isfile(src):
            os.rename(...)
```
Then the business logic only has to exist once

File: q2_cutadapt/_demux.py [hunk 2]
```diff
@@ -33,29 +34,61 @@ def run_command(cmd, verbose=True):
...
+            os.path.join(str(untrimmed_dir_fmt), 'reverse.fastq.gz'),
+            str(seqs_dir_fmt.forward_sequences.view(FastqGzFormat)),
+            str(seqs_dir_fmt.reverse_sequences.view(FastqGzFormat)),
+            ]
+    # SINGLE-END
+    else:
+        cmd = cmd + [
+            # {name} is a cutadapt convention for interpolating the sample id
+            # into the filename.
+            '-o', os.path.join(str(per_sample_dir_fmt), '{name}.fastq.gz'),
```
Comment (ebolyen) [inline q2_cutadapt/_demux.py]: If the name was always `{name}.1.fastq.gz` then you could share a bit more code between the single and paired-end cases (both in this helper, and in the rename).

File: q2_cutadapt/_demux.py (hunk 1 above)
Comment (ebolyen) [reply q2_cutadapt/_demux.py]: This only works if the initial filename is consistent between single and paired-end.

File: q2_cutadapt/tests/test_demux.py [hunk 3]
```diff
@@ -158,7 +161,62 @@ def test_variable_length_barcodes(self):
...
-class TestDemuxUtils(TestPluginBase):
+class TestDemuxPaired(TestPluginBase):
+    package = 'q2_cutadapt.tests'
+
+    def assert_demux_results(self, exp_samples_and_barcodes, obs_demuxed_art):
+        obs_demuxed = obs_demuxed_art.view(
+            SingleLanePerSamplePairedEndFastqDirFmt)
+        obs_demuxed_seqs = obs_demuxed.sequences.iter_views(FastqGzFormat)
+        # Since we are working with fwd/rev reads, duplicate each list elem
+        exp = [x for x in exp_samples_and_barcodes.iteritems() for _ in (0, 1)]
```
Comment (ebolyen) [inline q2_cutadapt/tests/test_demux.py]: I think this might be the same:
```python
exp = list(exp_samples_and_barcode.iteritems()) * 2
```

Comment (ebolyen) [reply q2_cutadapt/tests/test_demux.py]: nevermind. nested list comprehensions kill my brain every time I see them.

Comment (thermokarst) [reply q2_cutadapt/tests/test_demux.py]: I don't love the idea of the nested list comprehensions, but, for the sake of the test this was easy enough. I can rework to something more explicit if that would be helpful.

**Pull request merged**